### ISSUE TITLE
Grapheme clusters width miscalculation

### ISSUE THREAD

**Issue by haliphax** [Textualize/textual#4375]
It seems as though grapheme clusters (e.g. anything using skin tone modifiers, emoji variant selector, flags, etc.) are not properly accounted for when calculating the width of things. In the example program below, the scroll bar on initial load looks like a jagged mess. While it will smooth out after scrolling around a bit, a periodic refresh will break it again. I have seen some similar issues with placement of widgets on top of widgets containing grapheme clusters, as well (e.g. modal screens).

```python
from textual.app import App, ComposeResult
from textual.containers import VerticalScroll
from textual.widgets import Static


class ExampleApp(App):
    def compose(self) -> ComposeResult:
        yield VerticalScroll(Static())

    async def on_mount(self) -> None:
        widget = self.query_one(Static)
        widget.update("Grapheme clusters ✌️ 🧑‍🚒\n\n" * self.size.height)


if __name__ == "__main__":
    ExampleApp().run()
```

![image](https://github.com/Textualize/textual/assets/745888/ce246900-b1c5-46f4-96a8-4f7522b9ff42)

**Comment by haliphax** [comment]
Example with overlaying widgets:

![image](https://github.com/Textualize/textual/assets/745888/d2093ca1-42ad-4789-ac76-f0e752c3c733)

**Comment by haliphax** [comment]
Example with overlaying widgets:

![image](https://github.com/Textualize/textual/assets/745888/e656def3-d5fb-409b-923c-6100d042dc29)

**Comment by willmcgugan** [comment]
Yes, terminals render multi-codepoint emojis differently. And there is no way to know exactly how they will render
